### DISCUSSION TITLE
Manage App Store Connect API keys via fastlane match

### DISCUSSION THREAD

**Discussion by philprime** [fastlane/fastlane#19061] [Ideas]
Hi community,

this post is a combination of an idea and questions.

Due to a latest project, I have to manage multiple App Store Developer Accounts, and therefore their respective App Store Connect API key. I created a Fastfile script for myself, which uses the same logic as the internal match certificate management (encrypt/decrypt storage etc.). One lane allows me to add an API key to my encrypted git repository, and another to read it from there.

To be able to do so, I monkey-patched the internal encryption rules to also include the .p8 file.

Now I had the idea of expanding this proof-of-concept as a new command, e.g. `fastlane match add_api_key <path_to_p8_file>` or similar, as a Fork + Pull Request to this repository.

But before getting to work and investing the time, a few questions regarding this idea:

- Are there any security issues with this?

I have been using fastlane for many years, and I know the reasoning on why fastlane can be considered secure (as explained on [codesigning.guide](https://codesigning.guide/):

> Even if your certificates are leaked, they can't be used to cause any harm without your iTunes Connect login credentials. 

But my proposed would actually break with this concept, as the password to decrypt the storage, would also decrypt the access to the App Store Connect.

Additionally it breaks another common _law_, that credentials should never be committed to a repository. At least we could argue, that it is also encrypted with a passphrase.

- Is there any reason this hasn't been implemented yet (except for no-one had the time)? 

Before implementing my scripts, I looked for existing solutions, but couldn't find any ready-to-use ones. At some point I found a StackOverflow answer, explaining how something like this _could_ be achieved, but only on a high-level (will add a reference link when I find it again).

Looking forward to your feedback, and if you think it is worth the time to propose a PR and see how the maintainers react.

**Comment by philprime** [comment]
Bump?

**Comment by GoodMirek** [comment]
It will be indeed very helpful, I have the same use case.

**Comment by GoodMirek** [comment]
Looking through this Ideas discussion, I see that most questions are left unanswered for years ;-(

**Comment by philprime** [comment]
We eventually moved away from this approach, because it would break the security concepts behind fastlane match:

> Even if your certificates are leaked, they can't be used to cause any harm without your iTunes Connect login credentials

Ref: https://codesigning.guide/

**Reply by GoodMirek** [reply]
@philprime Thanks for your answer!

Is there any recommended approach for this use case?

**Reply by philprime** [reply]
After some evaluation we decided to go with a self-hosted HashiCorp Vault instance. 
Basically its main purpose is exactly what we were asking for: a vault to store the credentials.

You should read into its security features, such as encryption at rest.
https://www.vaultproject.io/
